### PR TITLE
Make replication stats return whole number

### PR DESCRIPTION
**Pull request by michalpristas**
**Description:** 

I failed to reproduce []uint8 to int64 conversion but I was able to repro float64 to int64 conversion error.
Different types may be due to different versions or values reported. 

The fix is forcing query to retrieve integer values. While this may seem like most obvious fix I'm not really aligned with it.

What query is returning for is a lag as a decimal number (whole part is seconds) by forcing this to return just an int we kind of losing precision. `0.4s` are reported as `0` while it is `400ms`. 

My proposal here would consists of 2 options.
First one is change reporting in a way that what we report is in fact time-span in `ms`. This could most likely be considered breaking.

Second option (I'm more in favor of) is to change the type of what is reported (from int to float). This way unit is intact and does not break possible visualizations, but we gain precision and won't lose data.

My first issue here so I wanted to get some feedback first before publishing something unreasonable.

_EDIT_

Went with the option of deprecating metrics with second precision (still fixing conversion failures) and introducing alternative to these metrics with `_ms` suffix in name and millisecond precision. 

Old metrics are now behind a featuregate which is enabled by default for now. 

**Link to tracking Issue:** #26714 

**Testing:** 
Setting up replicated postgres instances and testing method against this deployment.

**Documentation:** -

### CODE REVIEW

**Comment by djaglowski** [comment]
Thanks for the PR @michalpristas.

> Second option (I'm more in favor of) is to change the type of what is reported (from int to float). This way unit is intact and does not break possible visualizations, but we gain precision and won't lose data.

I think this is technically the right approach here, though obviously it would be a breaking change. The way we typically approach this is to use a featuregate to switch the behavior. One thing that would be simpler, if it makes sense, is to add new metrics and deprecate the old. I think we can only justify this if the new metric names are reasonable alternatives (e.g. not `foo.lag.old`.)

**Comment by michalpristas** [comment]
thanks for feedback, i will try to think about right approach

**Comment by michalpristas** [comment]
Went with the option of deprecating metrics with second precision (still fixing conversion failures) and introducing alternative to these metrics with _ms suffix in name and millisecond precision.

Old metrics are now behind a featuregate which is enabled by default for now.

also updated description of this PR with this info

**Review by djaglowski — COMMENTED** [review]
It doesn't appear this is actually adding a new metric, but rather changes the value assigned to the old one. You'll have to add the new metric in `metadata.yaml` and run `make generate`. 

Regarding the new metric name and unit - we should not include the unit in the metric name, since there is a dedicated field for defining the unit. Similarly, we should be able to return the new metric as a second but with greater precision, as opposed to using `ms`. Technically I think it is ok to use `ms`, but we typically lean towards `s` unless there is a strong reason not to e.g. the technology being monitored is expected to report a different unit.

**Comment by michalpristas** [comment]
ok, i think i understand how this works now. 
let me know if this is better. i added metrics with a different name, regenerated everything. I stayed with seconds as you said is more standard but added a precision by changing type to decimal.

File: receiver/postgresqlreceiver/metadata.yaml [hunk 1]
```diff
@@ -297,3 +297,13 @@ metrics:
       value_type: int
     extended_documentation: |
       This metric requires WAL to be enabled with at least one replica.
+  postgresql.wal.delay:
+    attributes: [wal_operation_lag, replication_client]
+    description: Time between flushing recent WAL locally and receiving notification that the standby server has completed an operation with it captured in seconds and milliseconds.
+    enabled: true
```
Comment (djaglowski) [inline receiver/postgresqlreceiver/metadata.yaml:303]: Unfortunately with new metrics we must make them optional, otherwise it is considered a breaking change.

File: .chloggen/replication-stats-conversion-failure.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: deprecation
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: postgresqlreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Deprecation of replication lag metrics in favor of more precise 'write_ms','replay_ms' and 'flush_ms'"
```
Comment (djaglowski) [inline .chloggen/replication-stats-conversion-failure.yaml:10]: Can we update this to specifically reference the metrics which is being deprecated (`postgresql.wal.lag`) and the one we're adding to replace it (`postgresql.wal.lag`)?

**Review by djaglowski — COMMENTED** [review]
Thanks for iterating @michalpristas.

I've approved CI. One thing I suspect is that you'll have to update one or more "golden" files in `testdata`, which should capture the expected output of the receiver.

File: receiver/postgresqlreceiver/metadata.yaml (hunk 1 above)
Comment (michalpristas) [reply receiver/postgresqlreceiver/metadata.yaml:303]: is switching `enabled` flag enough to make them optional?

Comment (djaglowski) [reply receiver/postgresqlreceiver/metadata.yaml]: Yes! I should have been clear on that. That's all we need, thanks.

**Comment by michalpristas** [comment]
I changed featureflag to stable and enabled feature.
It looks like Stable does not allow feature to be disabled and Deprecated does not allow feature to be enabled.
Usage of featuregate is a bit different now to what is described here in docs. Should I create an issue to update docs?

**Comment by djaglowski** [comment]
> Should I create an issue to update docs?

Please do!

My understanding of the deprecated state is that the feature flag itself is deprecated. For now, I think we need the flag to be alpha, which means it's disabled by default. This means we need to invert the usage of the gate so that enabling it removes the old metric.

**Comment by michalpristas** [comment]
ok this makes sense

File: receiver/hostmetricsreceiver/internal/scraper/loadscraper/load_scraper_windows.go [hunk 3]
```diff
@@ -140,7 +140,7 @@ func stopSampling(_ context.Context) error {
 		// no load scraper is running nothing to do
 		return nil
 	}
-	
+
```
Comment (djaglowski) [inline receiver/hostmetricsreceiver/internal/scraper/loadscraper/load_scraper_windows.go:143]: nit: This is unrelated - can it be removed from this PR? (Seems a reasonable change to submit separately though)

File: receiver/postgresqlreceiver/metadata.yaml [hunk 4]
```diff
@@ -297,3 +297,13 @@ metrics:
       value_type: int
     extended_documentation: |
       This metric requires WAL to be enabled with at least one replica.
+  postgresql.wal.delay:
+    attributes: [wal_operation_lag, replication_client]
+    description: Time between flushing recent WAL locally and receiving notification that the standby server has completed an operation with it captured in seconds and milliseconds.
```
Comment (djaglowski) [inline receiver/postgresqlreceiver/metadata.yaml:302]: We should leave units out of the description as well, since we have a dedicated field in the data model to indicate units.
```suggestion
    description: Time between flushing recent WAL locally and receiving notification that the standby server has completed an operation with it.
```

**Review by djaglowski — COMMENTED** [review]
Thanks for iterating @michalpristas. I've approved CI again.

One thing we will need is to include a scraper test with the gate enabled. Here's what I suggest:
- Create a new test by copying [`TestScraperNoDatabaseMultiple`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/040426fcd61a72988e263ee25e015f6bd7fdfb2d/receiver/postgresqlreceiver/scraper_test.go#L100) in `scraper_test.go` 
- Also copy the "expected" file used in the test and update the content to use match the new metric instead of the old.
- Use [`testutil.SetFeatureGateForTest`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/040426fcd61a72988e263ee25e015f6bd7fdfb2d/pkg/translator/prometheus/normalize_label_test.go#L16) to set the feature gate

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
Thanks for the improvement here @michalpristas.

We can bump the gate to beta as soon as +1 release from now, so roughly 2+ weeks.
